### DISCUSSION TITLE
Struggling with MultiOutput, single output works

### DISCUSSION THREAD

**Discussion by toblum** [pschatzmann/arduino-audio-tools#405] [Q&A] [answered]
Dear @pschatzmann,

I really like your library and it allowed me to play a stream or URL on a I2S board using the ESP32 in almost no time. That's great! But it cost me two whole evenings to make MultiOutput work, with no success yet. Before i give up, I thought I ask here.
My basic plan was to play a URLStream and flash a LED (or something else) according to the volume. I didn't find any informations on how to use VolumePrint class, so I went with FFT where I found an example.

My results so far:
- URLstream --> CSVStream = Works, get results in the console
- URLstream --> I2SStream = Works, can hear it
- URLstream --> AudioRealFFT = Works, get results in the console
- URLstream --> MultiOutput(I2SStream, CSVStream) = // Reads remote URL, doesn't start playing
- URLstream --> MultiOutput(I2SStream, AudioRealFFT) = // Reads remote URL, doesn't start playing

I think I'm just missing something little, but couldn't figure it out. I hope you can help me here, getting back on track.
I boiled down my sketch to the one that follows here.

Any help would be appreciated. Thanks!
Tobias

```
#include <Arduino.h>
#include "credentials.h"
#include "AudioTools.h"
#include "AudioLibs/AudioRealFFT.h" // or AudioKissFFT

int sampleRate = 16000;
int bitsPerSample = 16;
int channels = 1;

AudioRealFFT fft; // or AudioKissFFT
void fftResult(AudioFFTBase &fft)
{
	int diff;
	auto result = fft.result();
	if (result.magnitude > 100)
	{
		Serial.print(result.frequency);
		Serial.print(" ");
		Serial.print(result.magnitude);
		Serial.print(" => ");
		Serial.print(result.frequencyAsNote(diff));
		Serial.print(" diff: ");
		Serial.println(diff);
	}
}

URLStream urlStream(wifiUsername, wifiPassword);
I2SStream i2s;
WAVDecoder decoder(i2s);

EncodedAudioStream out2dec(i2s, decoder);
MultiOutput multiOut(fft, out2dec);

// StreamCopy copier(out2dec, urlStream); // Works
// StreamCopy copier(fft, urlStream); // Works
StreamCopy copier(multiOut, urlStream); // Reads URL, doesn't start playing/copying

void setup()
{
	Serial.begin(115200);
	AudioLogger::instance().begin(Serial, AudioLogger::Info);

	urlStream.begin("http://192.168.0.110:8080/quote_dl");

	// setup output - MAX98357A
	auto cfg = i2s.defaultConfig(TX_MODE);
	cfg.pin_ws = A17;
	cfg.pin_bck = A19;
	cfg.pin_data = A18;
	cfg.sample_rate = sampleRate;
	cfg.bits_per_sample = bitsPerSample;
	cfg.channels = channels;
	cfg.i2s_format = I2S_STD_FORMAT;
	i2s.begin(cfg);

	// Setup FFT
	auto tcfg = fft.defaultConfig();
	tcfg.length = 1024;
	tcfg.channels = channels;
	tcfg.sample_rate = sampleRate;
	tcfg.bits_per_sample = bitsPerSample;
	tcfg.callback = &fftResult;
	fft.begin(tcfg);
}

bool lastCopyState = false;
void loop()
{
	bool copyState = copier.copy();
	if (lastCopyState != copyState)
	{
		if (copyState) {
			Serial.println("copying");
		} else {
			Serial.println("NOT copying");
		}
		lastCopyState = copyState;
	}
}
```

**Comment by pschatzmann** [comment]
Not sure what exactly the issue is, but I see two sources of trouble:
1) Do the split after the decoding. You can't process any undecoded data (e.g. via fft). 
2) Timing: Make sure that the speed of all your processed methods is not providing data below the audio rate, otherwise I2s will not get the data fast enough - or in other words minimize the timing impact of what you do: Try to minimize the fft size and above all do not use prints!

**Comment by pschatzmann** [comment, marked as answer]
The method availableForPrint was misspelled in MultiOutput. I  just committed a correction.
W/o this correction you would need to call setCheckAvailableForWrite(false) on the copier

**Comment by toblum** [comment]
@pschatzmann Thank you for your quick and helpful support. I just tried it with the latest version and it works like a charm.
